### PR TITLE
CALCITE-2503 AvaticaCommonsHttpClientImpl client needs to set user-to…

### PR DESCRIPTION
**Pull request by karanmehta93**
…ken on HttpClientContext before sending the request

@joshelser @julianhyde Review please.

### CODE REVIEW

**Comment by joshelser** [comment]
Straightforward enough. How'd you catch this happening? Any thoughts about how we could test that this doesn't happen in the future?

**Comment by joshelser** [comment]
Nice, thanks for sharing those details. I don't think I did any perf testing when I was wiring up the SPNEGO authn, so that would explain how this slipped by.

> Any advice/suggestions appreciated.

Yeah, I don't have any good suggestions either. If tests are passing, that's good enough for me this time.

**Comment by joshelser** [comment]
> Any advice/suggestions appreciated.

Probably trying to handle this at a higher level would be easiest. I had written up a little test harness for wire compatibility testing. That could be expanded to also include some perf testing (coarse results, of course). Something to keep in the back of our minds, I think..

File: core/src/main/java/org/apache/calcite/avatica/remote/AvaticaCommonsHttpClientImpl.java [hunk 1]
```diff
@@ -215,7 +220,9 @@ HostnameVerifier getHostnameVerifier(HostnameVerification verification) {
         final int statusCode = response.getStatusLine().getStatusCode();
         if (HttpURLConnection.HTTP_OK == statusCode
             || HttpURLConnection.HTTP_INTERNAL_ERROR == statusCode) {
-          return EntityUtils.toByteArray(response.getEntity());
+          byte[] bytes = EntityUtils.toByteArray(response.getEntity());
```
Comment (joshelser) [inline core/src/main/java/org/apache/calcite/avatica/remote/AvaticaCommonsHttpClientImpl.java]: Is this order-dependent? I'm wondering if there's a reason you didn't change this to be:

```java
userToken = context.getUserToken();
return EntityUtils.toByteArray(response.getEntity());
```

Comment (karanmehta93) [reply core/src/main/java/org/apache/calcite/avatica/remote/AvaticaCommonsHttpClientImpl.java]: Good catch @joshelser 
Yes, that works too. Let me update the PR.
Somehow, I believed that the context can only be obtained after getting the entity, which is wrong.

**Comment by karanmehta93** [comment]
>  I don't think I did any perf testing when I was wiring up the SPNEGO authn, so that would explain how this slipped by.

If you were using TLS and not MTLS, you would not still find the issue. This is a really corner case and very hard to get into unless we have a specific use case :)

> Yeah, I don't have any good suggestions either. If tests are passing, that's good enough for me this time

Cool. Thanks for the review! Updated the PR as well

> That could be expanded to also include some perf testing (coarse results, of course). Something to keep in the back of our minds, I think.

Sure. Will do. Will be good if you can point me to that.

**Comment by joshelser** [comment]
> Sure. Will do. Will be good if you can point me to that.

https://github.com/apache/calcite-avatica/tree/master/tck probably needs some love already, but it's a start.

> If you were using TLS and not MTLS, you would not still find the issue. This is a really corner case and very hard to get into unless we have a specific use case :)

Ah, I was assuming that this would affect all cases with authn. Missed that it was only for MTLS :)

**Comment by joshelser** [comment]
```
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.724 s <<< FAILURE! - in org.apache.calcite.avatica.SslDriverTest
[ERROR] testReadWrite[0](org.apache.calcite.avatica.SslDriverTest)  Time elapsed: 0.082 s  <<< ERROR!
java.lang.RuntimeException: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
	at org.apache.calcite.avatica.SslDriverTest.testReadWrite(SslDriverTest.java:137)
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
	at org.apache.calcite.avatica.SslDriverTest.testReadWrite(SslDriverTest.java:137)
[ERROR] testReadWrite[1](org.apache.calcite.avatica.SslDriverTest)  Time elapsed: 0.037 s  <<< ERROR!
java.lang.RuntimeException: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
	at org.apache.calcite.avatica.SslDriverTest.testReadWrite(SslDriverTest.java:137)
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
	at org.apache.calcite.avatica.SslDriverTest.testReadWrite(SslDriverTest.java:137)
```

Seems like there are some issues with IBM Java, @karanmehta93. I'm not positive if these tests have been failing, or if they're newly introduced by your changes. Can you see?

**Comment by karanmehta93** [comment]
@joshelser I dont see IBM Java configured for master branch builds.
Sample: https://travis-ci.org/apache/calcite/builds/422274989

**Comment by joshelser** [comment]
@karanmehta93  https://travis-ci.org/apache/calcite-avatica/jobs/422237837

**Comment by karanmehta93** [comment]
I will try to repro it locally and fix it.

**Comment by joshelser** [comment]
Thanks, @karanmehta93. I just tested using the following and it reproduced the failure.

```
Java(TM) SE Runtime Environment (build 8.0.5.0 - pxa6480sr5-20170905_01(SR5))
IBM J9 VM (build 2.9, JRE 1.8.0 Linux amd64-64 Compressed References 20170901_363591 (JIT enabled, AOT enabled)
```

Hopefully you have a linux box you can use as IBM apparently doesn't provide JDK builds for OSX x.x

**Comment by joshelser** [comment]
Hrm, I'll have to look again. I had thought this wasn't failing before your change. Will take a look.

**Comment by joshelser** [comment]
> Hrm, I'll have to look again. I had thought this wasn't failing before your change. Will take a look.

Nope, looks like you're right. Not related. Let me try to commit this later tonight.

**Comment by joshelser** [comment]
```
-    <jetty.version>9.2.19.v20160908</jetty.version>
+    <jetty.version>9.4.11.v20180605</jetty.version>
```

I'd guess that something changed in this Jetty upgrade that broke IBM jdk. Not sure.

**Comment by karanmehta93** [comment]
@joshelser Any update?

**Comment by joshelser** [comment]
Sorry, in training. Will circle.

**Comment by risdenk** [comment]
@joshelser FYI tracked down the IBM JDK failure in CALCITE-2568
